File: src/main/app/src/Nexagon.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Text, Pattern, Path, Hex, HexUtils} from 'react-hexgrid';


class Nexagon extends Component {
    static propTypes = {
        q: PropTypes.number.isRequired,
        r: PropTypes.number.isRequired,
        s: PropTypes.number.isRequired,
        fill: PropTypes.string,
        cellStyle: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.object
        ]),
        className: PropTypes.string,
        data: PropTypes.object,
        onClick: PropTypes.func,
        children: PropTypes.node
    };

    static contextTypes = {
        layout: PropTypes.object, // TODO Shape
        points: PropTypes.string
    };

    constructor(props, context) {
        super(props, context);
        const {q, r, s} = props;
        const {layout} = context;
        const hex = new Hex(q, r, s);
        const pixel = HexUtils.hexToPixel(hex, layout);
        this.state = {hex, pixel};
    }

    // TODO Refactor to reduce duplicate
    componentWillReceiveProps(nextProps) {
        const {q, r, s} = nextProps;
        const {layout} = this.context;
        const hex = new Hex(q, r, s);
        const pixel = HexUtils.hexToPixel(hex, layout);
        this.setState({hex, pixel});
    }

    onClick(e) {
        if (this.props.onClick) {
            this.props.onClick(e, this);
        }
    }

    render() {
        const {fill, cellStyle, className} = this.props;
        const {points} = this.context;
        const {hex, pixel} = this.state;
        const fillId = (fill) ? `url(#${fill})` : null;
        return (
            <g
                className={classNames('hexagon-group', className)}
                transform={`translate(${pixel.x}, ${pixel.y})`}
                onClick={e => this.onClick(e)}
            >
                <g className="hexagon">
                    <polygon points={points} fill={fillId} style={cellStyle}/>
                    {this.props.children}
                </g>
            </g>
        );
    }
}

export default Nexagon;

File: src/main/java/de/bonndan/nivio/output/map/Nexagon.java
package de.bonndan.nivio.output.map;


import j2html.tags.ContainerTag;
import j2html.tags.DomContent;

import java.awt.geom.Point2D;

class Nexagon extends Component {

    private final String id;
    private DomContent children;
    private final Point2D.Double pixel;
    private final String fill;
    private final String cellStyle;

    Nexagon(DomContent children, ItemMapItem itemMapItem, Point2D.Double pixel, String fill, String cellStyle) {
        this.children = children;
        this.fill = fill;
        this.cellStyle = cellStyle;
        this.pixel = pixel;
        this.id = itemMapItem.landscapeItem.getFullyQualifiedIdentifier().toString();
    }

    public DomContent render() {

        var fillId = (fill) != null ? "url(#" + fill + ")" : null;

        ContainerTag inner = SvgTagCreator.g(
                SvgTagCreator.circle()
                        .attr("id", this.id)
                        .attr("cx", 0)
                        .attr("cy", 0)
                        .attr("r", 40)
                        .attr("fill", fillId)
                        .attr("style", cellStyle),
                children)
                .attr("class", "hexagon");

        return SvgTagCreator.g(inner)
                .attr("class", "hexagon-group")
                .attr("transform", "translate(" + pixel.x + "," + pixel.y + ")");
    }


}
